### ISSUE TITLE
Refactor visualisations editor panel

### ISSUE THREAD

**Issue by janagombitova** [akvo/akvo-lumen#1573]
# Context
Bar charts, how they are set up, what steps users take first and then next and how the editor is organised. 

Original discovery Mural: https://app.mural.co/t/akvo8194/m/akvo8194/1478624065782/e57b07e9bdf28011e96522735e27baf0beb4ee51

# Problem
We have many individual issues currently present in Lumen around bar charts. These exist because we allow "everything" to support the user in the most generic way possible. However this creates confusion and sometimes bar charts that can be wrongly interpreted. 

Secondly, when creating a bar chart we should guide Jerome on what he should do first and then next based on how the elements in the editor are organised. Currently the editor is, however, chaotic and it is hard to see and understand the structure of it. For example: 
- for a aggregation type to show Jerome needs to select the x axis. But in the editor the 1st axis to select is y, followed by aggregation type that we cannot select, then the label of y and then x axis. The way the elements are organised is confusing. 
- The default aggregation is mean, what makes sense if you have x and y selected. But what about cases when you only what to see the count for the categories on x. You only select x and default should be count to see the bars and the counts. 
- the x axis label shows all the way on the button of the editor even after the legend title
- data filters are hidden and not well visible in the UI

## Current issues: 
https://github.com/akvo/akvo-lumen/issues/1055 - Count in bar charts 
https://github.com/akvo/akvo-lumen/issues/1477 - Issue with stacked bars 
https://github.com/akvo/akvo-lumen/issues/1094 - No data sorting

## Requests to expand the functionality 
https://github.com/akvo/akvo-lumen/issues/1395 - Percentage stacked bars 
https://github.com/akvo/akvo-lumen/issues/595 - Horizontal bar charts
https://github.com/akvo/akvo-lumen/issues/1016 - Info fields to the editor items

## Extra issues - might be relevant
https://github.com/akvo/akvo-lumen/issues/846
https://github.com/akvo/akvo-lumen/issues/823

## User feedback 

"I never knew that only seeing the counts for the categories was possible. I always thought you need to select both x and y axis, so for cases where we just wanted to show number of submissions by submitter, for example, we used pie - what is not good if you have more than 2 categories. But it is good to know this is possible. In this case, it makes sense to start with the x axis for bar charts Because you get to see the initial count of the values in the variable before starting various aggregations and adding y-axis metric column." (Molo)

"I usually start with the x axis, but in Lumen you need to select both y and x to see the chart being generated even if you only want to see counts for categories" (Ilyasse)

"As a side note I find the bar chart editor a bit confusing. I would suggest to default to aggregation method Count, and the selection for the “bucket column” would come first. Only if you pick an aggregation different from Count do you need to select a metric column. This would mean that the bar config editor would very similar to the pie chart editor. With one additional option (except the sub-bucket stuff): “Aggregation”. If you pick aggregation other than count, then you would see the metric column options. Conceptually, one way to see why the current setup is fishy, is that if you select the count aggregation the metric column selection is arbitrary. Whatever you select you get the same chart but you still have to select one" (Jonas)

# Solution or next step
*How could we solve it?*

Meet up to talk about how one goes about creating a bar chart, what elements depend on what and then decide how to organise the editor panel to support the workflow. We need to look at how the elements are organised currently and see if we can group them in a better way and to look at what elements come first and then next to support a better user workflow when creating bar charts.Then we need to break down the decision to small implementable issues.

**Comment by finnfiddle** [comment]
```
----------------
Type: Bar
----------------
dataset
----------------
x axis
  column - required - could be good first choice for user ***visualization appears***
  ---------- advance ---------
  title
  grouping
  limiting
----------------
y axis
  column
  aggregation type
    - count if y column not set and disabled
    - else if hasnt been manually set by user shows default: mean
  --------- advance ---------
  percentages/absolute values - is this only visible when stacked (grouped)
  title
----------------
sorting
----------------
misc
  orientation - vertical default
  split/stacked - only displayed if do grouping
  legend title
----------------

```

**Comment by Kiarii** [comment]
below is a sketch showcasing the compartmentalization of different sections in the visualization editor and the advanced setting expand-collapse. Others improvements suggested in the sketch could improve the UX, i.e. if we have the time

![viz editor sketch 2 2x](https://user-images.githubusercontent.com/12692444/44715112-a3e57700-aabf-11e8-90fc-d9d59ae7c3c9.png)

**Comment by janagombitova** [comment]
I have been testing the implementation on dark-test and the 1st impression I have is that this is such an amazing improvement, we can be proud! Clean, understandable, amazing. 

I have a concern about the part where we pick the visualisation type. We have 8 options and you need to scroll horizontally to see the option. Not all machines support horizontal scrolling so you need to see the scroll bar. My larger concern is that if you visualisation type is in the 2nd half of the list, you cannot see it is selected. 

Can we reconsider this approach? how about showing all options in a grid of two lines? Or another option?
  
![editor](https://user-images.githubusercontent.com/12456965/45424894-2e210400-b698-11e8-967f-5da94cec3104.gif)

**Comment by finnfiddle** [comment]
I can think of 2 options:
- putting them on 2 rows as you say @janagombitova 
- putting little arrows on the sides to let users click and scroll horizontally - and when the page loads ensuring that it scrolls to the selected one

**Comment by finnfiddle** [comment]
@Kiarii thoughts? ^

**Comment by janagombitova** [comment]
# Did editor changes change existing visualisations?
I checked randomly existing visualisations side by side on test and dark-test and could not find a difference. But it would be good if another pair of eyes confirmed this too. 

# Difference between old and new editor
In regards to the editor changes here are things that we miss in comparison to the live version now: 
- on maps, when creating a geoshape layer, the drop downs miss their descriptions thus you do not know what you are selecting for

<img width="1243" alt="screen shot 2018-09-12 at 17 14 02" src="https://user-images.githubusercontent.com/12456965/45441533-8ae3e500-b6bf-11e8-8183-3f8ea768411e.png">
<img width="1241" alt="screen shot 2018-09-12 at 17 15 30" src="https://user-images.githubusercontent.com/12456965/45441535-8ae3e500-b6bf-11e8-9864-0476e3a26ce6.png">

- for donut and pie chart we state y-axis what is a bug 

<img width="1254" alt="screen shot 2018-09-12 at 17 22 10" src="https://user-images.githubusercontent.com/12456965/45441602-b8c92980-b6bf-11e8-8f24-4393c1ab115c.png">

# Further improvements to the editor

I think it might be better to do the re-design as one large change and then perhaps the other improvements as separate changes. What do you think @kardan and @finnfiddle ? 

For this batch of changes I would then suggest we only address the horizontal scroll and ensure nothing is broken.

**Comment by Kiarii** [comment]
<img width="1265" alt="screen shot 2018-09-13 at 9 07 25" src="https://user-images.githubusercontent.com/12692444/45474275-04241c00-b742-11e8-9404-c05d92cb74f7.png">
On the above illustration; I think we should instead of showing text show the max number of bars/elements by default thereby cutting the user some slack; instead of leaving them to figure out what they would need to do and where to bring some bars in view. This would be mean that in this case we expand the 'Advanced' section of the X axis and show what options the system has taken on behalf of the user e.g.

<img width="255" alt="screen shot 2018-09-13 at 10 48 08" src="https://user-images.githubusercontent.com/12692444/45474473-8c0a2600-b742-11e8-8c33-76fe675c4a6c.png">

The note of why there are only 200 or 'max' elements can be added below the size so the user is aware of what they are looking at and why the system put a cap on the # of those

**Comment by janagombitova** [comment]
@Kiarii I agree with your comment above, but think that it touches on a  separate problem - How to handle bar charts with more than 200 element - so it should be handled separately so we do not push too many different changes into one issue. Can you recreate the comment into a new issue?

**Comment by finnfiddle** [comment]
##TODO

1. [ ] permanent horizontal scroll vis type + scroll to selected on load
2. [ ] blue radio buttons
3. [ ] unify how we call y and x axis as for some visualisations we call one 'metric' and the other 'bucket' or both 'metric' but the new design does not reflect that
4. [ ] on maps, simplify navigation between layers (if possible) and now the back button takes quite some space
5. [ ] on maps, when creating a geoshape layer, the drop downs miss their descriptions thus you do not know what you are selecting for
6. [ ] for donut and pie chart we state y-axis what is a bug
7. [ ] in pivot tables the warning for aggregation does not fully show
8. [ ] in bar charts I would not hide the Aggregation under Advanced as it is an important piece of information on how the chart is built
9. [ ] On the above illustration; I think we should instead of showing text show the max number of bars/elements by default thereby cutting the user some slack; instead of leaving them to figure out what they would need to do and where to bring some bars in view. This would be mean that in this case we expand the 'Advanced' section of the X axis and show what options the system has taken on behalf of the user e.g.
10. [ ] The note of why there are only 200 or 'max' elements can be added below the size so the user is aware of what they are looking at and why the system put a cap on the # of those
11. [ ] It would also be nice to standardize the toggle action, placement on the right makes more sense if we think of users reading left to right; you read and then take action.. see these two
12. [ ] I also think on the first section of the visualization, it is enough to have "Type" in place of "Visualization type"
13. [ ] If we have the time, I would also like us to take a closer look at the 'advanced' settings in Y and especially X; break them down and make them more visual e.g.

**Comment by finnfiddle** [comment]
@Kiarii what do you think of using the blue instead of the lumen green? To me blue is a neutral colour. Green indicates something positive or successful in the ui - like a correct validation or alert. Similarly to red indicating an error.

**Comment by Kiarii** [comment]
@janagombitova let me takeup this with Loic since he and Linda were doing the visual design for lumen, then I will get back to you

**Comment by Kiarii** [comment]
![image](https://user-images.githubusercontent.com/12692444/45612821-e4a43080-ba6c-11e8-82e3-e04321c21800.png)

On the "Back to Layers" however, I think a back arrow should suffice without texts, and this shd be placed on the same line as we had initially; unless we can confirm that this creates a confusion with users..

**Comment by finnfiddle** [comment]
@Kiarii any ideas for this one "on maps, simplify navigation between layers (if possible) and now the back button takes quite some space"?
<img width="419" alt="screen shot 2018-09-17 at 12 35 20 pm" src="https://user-images.githubusercontent.com/4628721/45615773-2d141c00-ba76-11e8-99e9-80d90b9cab9d.png">
<img width="421" alt="screen shot 2018-09-17 at 12 35 06 pm 1" src="https://user-images.githubusercontent.com/4628721/45615775-2d141c00-ba76-11e8-8460-4680083dfa03.png">

**Comment by Kiarii** [comment]
@finnfiddle let me have a think on possibilities to simplify navigation within the layers..

**Comment by finnfiddle** [comment]
@Kiarii @janagombitova what should we do with this task? Are we happy with the sidebars now or we need another iteration? If so new ticket or keep this one?

**Comment by finnfiddle** [comment]
@Kiarii its easier said than done. There is no cross browser solution to always show the scrollbar. So I added the blue bar as an indicator. Reimplementing custom scrollbars would have been a bigger job. This already took me quite a while and was deceptively challenging to implement

**Comment by Kiarii** [comment]
I see

**Comment by janagombitova** [comment]
@finnfiddle and @Kiarii I think we will definitely need another iteration, but we need to learn first. We should now wait for feedback from actual users on their experience with the new design. Once we have concrete improvements those should go into small separate issues. So I think this one can be closed with the release.
